parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveType.java

```diff
@@ -514,16 +514,41 @@ protected Type union(Type toMerge) {
     return union(toMerge, true);
   }
 
+  private void reportSchemaMergeError(Type toMerge) {
+    throw new IncompatibleSchemaModificationException("can not merge type " + toMerge + " into " + this);
+  }
+
   @Override
   protected Type union(Type toMerge, boolean strict) {
-    if (!toMerge.isPrimitive() || (strict && !primitive.equals(toMerge.asPrimitiveType().getPrimitiveTypeName()))) {
-      throw new IncompatibleSchemaModificationException("can not merge type " + toMerge + " into " + this);
+    if (!toMerge.isPrimitive()) {
+      reportSchemaMergeError(toMerge);
+    }
+
+    if (strict) {
+      // Can't merge primitive fields of different type names
+      if (!primitive.equals(toMerge.asPrimitiveType().getPrimitiveTypeName())) {
+        reportSchemaMergeError(toMerge);
+      }
+
+      // Can't merge primitive fields of different original types
+      if (getOriginalType() == null && toMerge.getOriginalType() != null ||
+          getOriginalType() != null && !getOriginalType().equals(toMerge.getOriginalType())) {
+        reportSchemaMergeError(toMerge);
```

Comment:
Thanks! Being in the Scala world for so long a time, I really need to re-learn Java again :)